proposals/0000-mutable-fields.rst

```diff
@@ -0,0 +1,473 @@
+.. proposal-number::
+
+.. trac-ticket::
+
+.. implemented::
+
+.. sectnum::
+
+.. contents::
+
+Mutable Constructor Fields
+==========================
+
+We propose an extension to add support for mutable constructor fields.
+
+Motivation
+----------
+
+This proposal improves both the runtime and memory overhead for code
+that currently uses ``IORef`` or ``STRef``.
+
+The best representation we currently have available for a type that
+contains an ``IORef`` is something like::
+
+  data MutPair a b =
+    MutPair {-# UNPACK #-} !(IORef a)
+            {-# UNPACK #-} !(IORef b)
+
+and the runtime representation of this is 3 heap objects: the outer
+``MutPair`` with 2 pointer fields (3 words total), and 2 ``MutVar#``
+objects (2 words each).
+
+Under this proposal, we would be able to declare a ``MutPair`` that is
+represented by a single 3-word heap object containing 2 directly
+mutable pointer fields.
+
+Removing a layer of indirection can make a big difference.  Currently
+``IORef`` consists of two objects: the outer ``IORef`` constructor and
+the inner ``MutVar#`` type.  This proposal allows the two layers to be
+collapsed. Operations on the collapsed ``IORef`` are approximately 2x
+faster than the current ``IORef`` (according to `rough
+experiments <https://gist.github.com/simonmar/d8a05797c01799abcb979aacf27164c7>`_).
+Note that the outer constructor in the current ``IORef``
+representation can be eliminated by using ``UNPACK`` or
+sometimes automatically by strictness analysis, but not always.  The proposed
+representation is *always* cheap.
+
+This extension is aimed at particularly performance-critical data
+structures, where the library writer is prepared to invest some effort
+into defining an optimal representation.  This extension makes it
+possible to eliminate a layer of indirection in mutable data
+structures in a type-safe way, where currently this can only be done
+by using ``unsafeCoerce`` tricks; see for example `the structs
+package <http://hackage.haskell.org/package/structs>`_.
+
+As a special case, this proposal also provides for the declaration of
+constructors with *identity*. That is, constructors that are created
+by an explicitly effectful operation, and that can be compared using
+pointer-equality.  Currently only a few built-in types like ``IORef``
```

Comment:
Does the pointer equality need to be a primitive, or can it be implemented with a magic type class and `reallyUnsafePtrEquality#`?